File: Data Structure/python/linked_list.py
       
class MyLinkedList:
    def __init__(self, value):
        self.head = {
            "value" : value,
            "next" : None
        }
        self.tail = self.head
        self.length = 1

    def append(self, value):
        newNode = {
            "value": value,
            "next": None
        }
        self.tail["next"] = newNode
        self.tail = newNode
        self.length += 1
        return self

    def prepend(self, value):
        newNode = {
            "value": value,
            "next": None
        }
        newNode["next"] = self.head
        self.head = newNode
        self.length += 1
        return self

    def insert(self, index, value):
        if(index >= self.length):
            return self.append(value)
        
        newNode = {
            "value" : value,
            "next" : None
        }
        leader = self.traverseToIndex(index-1)
        holdingPointer = leader["next"]
        leader["next"] = newNode
        newNode["next"] = holdingPointer
        self.length += 1
        
    
    def traverseToIndex(self,index):
        # check params
        counter = 0
        currentNode = self.head
        while(counter != index):
            currentNode = currentNode["next"]
            counter += 1
        
        return currentNode



newList = MyLinkedList(10)
# newList.prepend(15)
# newList.prepend(30)
newList.append(20)
newList.append(5)
# newList.insert(3,17)
# newList.insert(5,24)
# newList.insert(9, 65)
print(newList.head)
print(newList.length)

File: Data_Structure/try/linkedlist.js
class Node{
    constructor(value){
            this.value = value,
            this.next = null
    }
}

class LinkedList{
    constructor(value){
        this.length = 1
        this.head = {
            value : value,
            next : null
        }
        this.tail = this.head
    }

    append(value){
        const node = new Node(value)
        this.tail.next = node
        this.tail = node
        this.length++
    }

    prepend(value){
        const node = new Node(value)
        node.next = this.head
        this.head = node
        this.length++
    }

    printList(){
        const array = []
        let currentNode = this.head
        while(currentNode !== null){
            array.push(currentNode.value)
            currentNode = currentNode.next
        }
        return array;
    }

    addNode(index, value){
        const node = new Node(value)
        if(index >= this.length){
            return this.append(value)
        }
        const ladder = this.tranverseToIndex(index-1)
        const holdingPointer = ladder.next
        ladder.next = node
        node.next = holdingPointer
        this.length++
        return this.printList();
    }

    tranverseToIndex(index){
        let counter = 0;
        let currentNode = this.head;
        while(counter !== index){
            currentNode = currentNode.next
            counter++
        }
        return currentNode;
    }

    remove(index){
        const ladder = this.tranverseToIndex(index-1)
        const unwantedNode = ladder.next
        ladder.next = unwantedNode.next
        this.length--
        return this.printList()
    }
}


const newList = new LinkedList(10)
// newList.prepend(2)
// newList.prepend(4)
newList.append(15)
newList.append(20)
newList.append(25)
newList.append(30)
newList.addNode(10, "a")
newList.addNode(10, "b")
newList.remove(4)
console.log(newList.printList())